lib/shopify_cli/command.rb

```diff
@@ -58,6 +63,62 @@ def subcommand_registry
         )
       end
 
+      def recommend_ruby(from:, to:)
+        @compatible_ruby_range = VersionRange.new(
+          from: Semantic::Version.new(from),
+          to: Semantic::Version.new(to)
+        )
+      end
+
+      def recommend_default_ruby_range
+        recommend_ruby(
+          from: Constants::SupportedVersions::Ruby::FROM,
+          to: Constants::SupportedVersions::Ruby::TO
+        )
+      end
+
+      def check_ruby_version
+        check_version(
+          Environment.ruby_version,
+          range: @compatible_ruby_range,
+          runtime: "Ruby"
+        )
+      end
+
+      def recommend_node(from:, to:)
+        @compatible_node_range = VersionRange.new(
+          from: Semantic::Version.new(from),
+          to: Semantic::Version.new(to)
+        )
+      end
+
+      def recommend_default_node_range
+        recommend_node(
+          from: Constants::SupportedVersions::Node::FROM,
+          to: Constants::SupportedVersions::Node::TO
+        )
+      end
+
+      def check_node_version
+        check_version(
+          Environment.node_version,
+          range: @compatible_node_range,
+          runtime: "Node"
+        )
+      end
+
+      def check_version(version, range:, runtime:)
+        return if Environment.test?
+        return if range.nil?
+        return if version >= range.from && version < range.to
+        Context.new.warn("Your environment #{runtime} version, #{version},"\
```

Comment:
@MeredithCastile I need your help with this content. Here's screenshot of how it looks currently:

![image](https://user-images.githubusercontent.com/663605/150122371-d2cd09b4-04b6-4f97-bbe7-df6678f2dfd3.png)

Reply:
@pepicrft @amcaplan Lmk if this works. Thx!

Reply:
@MeredithCastile  I think we'll need to improve the error in a future release, as this has been merged already...

To answer questions as best I can:

1. DISCOVERABILITY. This is implemented as a hook which will run before any potentially impacted CLI action.
2. RECOMMENDED ACTION. Good point, I think the current version assumes developers will know how to change it but we should be more proactive about giving guidance. It's a bit complicated, as version management is hairy for Ruby (there are 3 major version managers: RVM, rbenv, chruby) and many developers aren't really Ruby developers anyway so they're just using Mac system Ruby. Needs some more thought. Node is more simple as probably most users are using `nvm`, but there are also `n`, `fnm`, and probably many smaller competitors because JS gonna JS. Perhaps we provide an example for each of the most popular version managers, or link to a docs page?
3. MAINTENANCE. The `FROM` is a minimum version which might be updated at some point but we'll know to do that, I think. The `TO` version, following some discussion today, will be more broadly scoped, so rather than keeping it pointing to the latest patch version, it'll be pointing to the next minor version we don't support (for Ruby) which is an annual event. For Node, I assume we'll adopt a similar approach, specifying the first release we don't expect to support. So when we support Node 17 fully, we'll bump it to 18 that we don't support.